src/worker/typosquat/database.rs

```diff
@@ -0,0 +1,230 @@
+        // term.
+        for result in crate_owners::table
+            .filter(crate_owners::deleted.eq(false))
+            .filter(crate_owners::crate_id.eq_any(crates.keys().collect::<Vec<_>>()))
+            .select((
+                crate_owners::crate_id,
+                crate_owners::owner_id,
+                crate_owners::owner_kind,
+            ))
+            .load_iter::<(i32, i32, i32), DefaultLoadingMode>(conn)?
+        {
+            let (crate_id, owner_id, owner_kind) = result?;
+            crates.entry(crate_id).and_modify(|(_name, krate)| {
+                krate.owners.insert(Owner::new(owner_id, owner_kind));
+            });
+        }
+
+        Ok(Self {
+            crates: crates.into_values().collect(),
+        })
+    }
+}
+
+impl Corpus for TopCrates {
+    fn contains_name(&self, name: &str) -> typomania::Result<bool> {
+        Ok(self.crates.contains_key(name))
+    }
+
+    fn get(&self, name: &str) -> typomania::Result<Option<&dyn Package>> {
+        Ok(self.crates.get(name).map(|krate| krate as &dyn Package))
+    }
+}
+
+pub struct Crate {
+    owners: HashSet<Owner>,
+}
+
+impl Crate {
+    /// Hydrates a crate and its owners from the database given the crate name.
+    pub fn from_name(conn: &mut PgConnection, name: &str) -> QueryResult<Self> {
+        use crate::models;
+        use diesel::prelude::*;
+
+        let krate = models::Crate::by_exact_name(name).first(conn)?;
+        let owners = krate.owners(conn)?.into_iter().map(Owner::from).collect();
+
+        Ok(Self { owners })
+    }
+}
+
+impl Package for Crate {
+    fn authors(&self) -> &dyn typomania::AuthorSet {
+        self
+    }
+
+    fn description(&self) -> Option<&str> {
+        // We don't do any checks that require descriptions.
+        None
+    }
+
+    fn shared_authors(&self, other: &dyn typomania::AuthorSet) -> bool {
+        self.owners
+            .iter()
+            .any(|owner| other.contains(owner.borrow()))
+    }
+}
+
+impl AuthorSet for Crate {
+    fn contains(&self, author: &str) -> bool {
+        self.owners.contains(author)
+    }
+}
+
+/// A representation of an individual owner that can be compared to other owners to determine if
+/// they represent the same unique user or team that may own one or more crates.
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+struct Owner(String);
+
+impl Owner {
+    fn new(id: i32, kind: i32) -> Self {
+        Self(format!("{kind}::{id}"))
+    }
+}
+
+impl Borrow<str> for Owner {
+    fn borrow(&self) -> &str {
+        &self.0
+    }
+}
+
+impl From<crate::models::Owner> for Owner {
+    fn from(value: crate::models::Owner) -> Self {
+        Self::new(value.id(), value.kind())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{test_util::pg_connection, worker::typosquat::test_util::Faker};
+    use thiserror::Error;
+
+    use super::*;
+
+    #[test]
+    fn top_crates() -> Result<(), Error> {
```

Comment:
wouldn't it be easier to use `anyhow::Result` directly instead of the custom error type? seems like we don't really care about any of the specific enum variants?

Reply:
`typomania::Result` doesn't require an error that can coerce to `anyhow::Error` (because it doesn't include `Sync` as a constraint). That's probably a mistake, but applying a little bit of `thiserror` was the quickest way around the API issue for now.

Reply:
hmm, okay. guess we can clean this up later.